src/Haskell/Ide/Engine/Types.hs

```diff
@@ -32,6 +32,6 @@ data GhcRequest = forall a. GhcRequest
 
 data IdeRequest = forall a. IdeRequest
   { pureReqId :: J.LspId
-  , pureReqCallback :: a -> IO ()
-  , pureReq :: IdeM a
+  , pureReqCallback :: IdeResponse a -> IO ()
```

Comment:
I think it would be a good idea to have the type of the callback be `a -> IO ()` and we let the dispatcher handle error/deferred responses.